packages/lodestar/src/sync/sync.ts

```diff
@@ -197,6 +197,11 @@ export class BeaconSync implements IBeaconSync {
   }
 
   private onUnknownBlockRoot = async (root: Root): Promise<void> => {
+    // at regular sync time, we fetch follow-up range so don't need to handle this
+    // otherwise, got "write EPIPE" error bc we call beacon_block_by_root too much
+    // at the same time to same peer and it disconnects us
+    // this happens with "missing parent" gossip blocks
+    if (!this.isSynced()) return;
```

Comment:
but if we are on `current slot - 2` and receive gossip block at `current slot - 2` we would need that to fetch missing parent? Or are you hopping we would fetch range that would include missing block?

Side question after regular sync is completed, what service handles unknown block root events?

Reply:
> but if we are on current slot - 2 and receive gossip block at current slot - 2 we would need that to fetch missing parent? Or are you hopping we would fetch range that would include missing block?

we should have gossip parent block at that time bc we fetch ranges from behind until missing parent gossip blocks, given the fact that we use single peer for regular sync so it should return the same chain to head

> Side question after regular sync is completed, what service handles unknown block root events?

this `sync.ts` still handles "unknownBlockRoot" event after the node is synced. I think it not only manages initial sync/regular sync but also after that time. Wdyt?
One possible fix for this issue could be only register `unknownBlockRoot` event after regular sync completed.

Reply:
> we should have gossip parent block at that time bc we fetch ranges from behind until missing parent gossip blocks, given the fact that we use single peer for regular sync so it should return the same chain to head

I'm worried because we are syncing from one peer which might not have latest head at the time we need it. I agree there is low chance for that but it could still stall us I guess.

> One possible fix for this issue could be only register unknownBlockRoot event after regular sync completed.

But we kinda need handling that before and during sync as peers could return range with some block missing (it happens quite often actually).

Reply:
> I'm worried because we are syncing from one peer which might not have latest head at the time we need it. I agree there is low chance for that but it could still stall us I guess.

I found that "missing parent" gossip blocks come regularly so as long as the best peer does not way below head, we're good since "follow up" range and "missing parent gossip blocks" range will gradually overlap and we get through regular sync

> But we kinda need handling that before and during sync as peers could return range with some block missing (it happens quite often actually).

Agree. So this main `sync.ts` service takes care "after sync", and the regular sync should handle "unknowBlockRoot" event itself (will do it in an upcoming PR for regular sync).